### DISCUSSION TITLE
Default connection for migrations

### DISCUSSION THREAD

**Discussion by aagjalpankaj** [laravel/framework#54035] [Ideas]
Currently Laravel provides two ways to specify database connection for migrations:
1. In the migrate command e.g. `php artisan migrate --database=<connection_name>`
2. In the migration itself e.g. `Schema::connection('connection_name')->`

Otherwise it uses the default connection specified in `database.php`.

Can we allow specifying default connection just for migrations like below in `database.php`?

```
'migrations' => [
        'table' => 'migrations',
        'update_date_on_publish' => true,
        'connection' => env('DB_MIGRATION_CONNECTION'),
    ],
```

---
### What prompts us to trigger this requirement?
We create two separate database connections - one for application and another for migration. Only migration user has DDL permissions. We use [laravel automations](https://serversideup.net/open-source/docker-php/docs/laravel/laravel-automations) and that does't allow `--database` option while executing migrations. So we always need to specify connection explicitly in the migration using `Schema::connection('connection_name')->`.

I understand that this requirement may seem superfluous and not urgent, but good to see what you guys think about it :)

**Comment by Sophist-UK** [comment]
My suggestion would be for you to sub-class the `Illuminate\Database\Migrations\Migration` class such that your sub-class always uses the `Schema:connection` call.

**Reply by aagjalpankaj** [reply]
Yes, there are workarounds to achieve it. default option in configuration would be nice to have :)

**Reply by Sophist-UK** [reply]
My suggestion is ***not*** a workaround - it is a solution.

This requirement is NOT a common one - and on that basis is extremely unlikely to be adopted in core Laravel, especially when the above solution is a reasonable one and an easy one and one which fits entirely within the concepts of OO.

**Reply by kokoshneta** [reply]
> This requirement is NOT a common one

Strongly disagree. As a base rule of thumb, the user your application uses to connect to your database for normal CRUD operations (usually the default connection in your database file) should only have the privileges it needs. Altering and deleting table structures is only required for migrations and should not be available to the app user at all; having a default connection for your app and an elevated connection for migrations should be the default.

**Reply by Sophist-UK** [reply]
I agree that having a separate set of credentials for normal CRUD and for management operations is good security practice in production, but it would require more than simply having two connections in the ENV file to achieve this - even putting the management credentials into a secret wouldn't help because it would still be accessible to someone who has hacked the code.

This would actually need specific code inside Laravel itself to distinguish between CRUD-only and Management connections.

**Reply by aagjalpankaj** [reply]
This is how I am expecting:
![image](https://github.com/user-attachments/assets/d7261374-4b16-4d21-82da-66b6974ae49f)

**Reply by Sophist-UK** [reply]
Well, you can add a second connection to the env/config, and you can sub-class the `Illuminate\Database\Migrations\Migration` class so that migrations use the alternate connection credentials.

**Reply by aagjalpankaj** [reply]
That I can achieve by specifying connection in the migration itself as mentioned in the description. e.g. `Schema::connection('connection_name')->`.
How will you suggest the creating `migrations` table using `migration` connection without specifying `--database` option in the migrate command? The `artisan migrate` uses default connection for creating `migrations` table.

**Reply by kokoshneta** [reply]
> How will you suggest the creating `migrations` table using `migration` connection without specifying `--database` option in the migrate command? The `artisan migrate` uses default connection for creating `migrations` table.

That’s what you get by sub-classing the `Migration` class. What I’ve done is this:

- in your `config/database.php`, add a `migration` key after `default` and set it to whatever connection you want to use (e.g., `env('DB_MIGRATIONS_CONNECTION')` – assuming you have defined a migrations connection in your .env file)
- create a class `/app/Database/Migrations/Migration.php` (or wherever you prefer to put it) and have it extend `Illuminate\Database\Migrations\Migration`
- override the `getConnection()` method to make it return `config('database.migration')`
- do a global search for `use Illuminate\Database\Migrations\Migration;` and replace all with `use App\Database\Migrations\Migration;` (or wherever you put your class)

That will make sure all your existing migrations use the connection you want. If you want to automate it for future migrations, [publish your stubs](https://laravel.com/docs/10.x/artisan#stub-customization) and change the `use` statements in the three migration stubs as well.

**Reply by djibarian** [reply]
@kokoshneta as far as I understand you can achieve exactly the same adding `protected $connection = <conn>;` to your migration. All that stuff you mentioned only makes avoiding that line in every migration, but the `migrations` table will be still created using the connection defined in `database.default`, which is not what desired by @aagjalpankaj. So... how to determine where to create the `migrations` table if you want it to be created using a different connection than the default one?

**Reply by kokoshneta** [reply]
@djibarian That’s very true. The first time you run `artisan migrate`, you’ll have to specify the connection in the `--database` option manually to make sure the `migrations` table gets created properly. It is only a **workaround**, after all, and it would be far preferable for the framework to support simply specifying a migration connection in the config file.

A possible alternative way to do it just occurred to me, though I haven’t tried or tested and don’t know if it will actually work:

The [`migrate:install` command](https://github.com/laravel/framework/blob/9fe1f12e05745620969ef8132205660e563f322e/src/Illuminate/Database/Console/Migrations/InstallCommand.php#L39) expects a parameter that implements `MigrationRepositoryInterface`, which (as far as I can see) is resolved via the service container. Within the framework, that is the [`DatabaseMigrationRepository`](https://github.com/laravel/framework/blob/9fe1f12e05745620969ef8132205660e563f322e/src/Illuminate/Database/Migrations/DatabaseMigrationRepository.php#L8).

Instead of subclassing the `Migration` class, it might be easier to subclass the `DatabaseMigrationRepository` class and set the connection in the constructor:

```php
namespace App\Database\Migrations;

class DatabaseMigrationRepository extends \Illuminate\Database\Migrations\DatabaseMigrationRepository {
  public function __construct(Resolver $resolver, $table) {
    parent::__construct($resolver, $table);
    $this->resolver->setDefaultConnection('your-migration-connection');
  }
}
```

Then [bind the `MigrationRepositoryInterface` to your subclass](https://laravel.com/docs/12.x/container#binding-interfaces-to-implementations):

```php
// app/Providers/AppServiceProvider.php

class AppServiceProvider extends ServiceProvider {
  public function register() {
    $this->app->bind(\Illuminate\Database\Migrations\MigrationRepositoryInterface::class, \App\Database\Migrations\DatabaseMigrationRepository::class);
  }
}
```

Unless I’m missing something, that _should_ pull in the repository class with the correct connection everywhere that the application expects to do something migration-related.

**Comment by shaedrich** [comment]
No harm in having options here—go for it, I'd say 👍🏻

**Comment by aagjalpankaj** [comment]
Created PR: https://github.com/laravel/framework/pull/58944
